File: 547. Number of Provinces/main.py
from typing import List


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        def dfs(current: int):
            visited[current] = 1

            for j in range(len(isConnected[current])):
                if isConnected[current][j] == 1 and visited[j] == 0:
                    dfs(j)

        visited = [0] * len(isConnected)
        count = 0

        for i in range(len(isConnected)):
            if visited[i] == 0:
                dfs(i)
                count += 1

        return count


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        def find(x: int) -> int:
            if parent[x] != x:
                parent[x] = find(parent[x])

            return parent[x]

        def unionSet(x: int, y: int):
            rootX = find(x)
            rootY = find(y)

            if rootX != rootY:
                parent[rootX] = rootY

        parent = [i for i in range(len(isConnected))]

        for i in range(len(isConnected)):
            for j in range(len(isConnected[i])):
                if isConnected[i][j] == 1:
                    unionSet(i, j)

        count = 0

        for i in range(len(parent)):
            if parent[i] == i:
                count += 1

        return count

File: 547. Number of Provinces/main.cpp
#include <vector>

using namespace std;

class Solution {
    vector<int> parent;

    void init(int size) {
        parent.resize(size);

        for (int i = 0; i < size; i++) {
            parent[i] = i;
        }
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }

        return parent[x];
    }

    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            parent[rootX] = rootY;
        }
    }

public:
    int findCircleNum(vector<vector<int>> &isConnected) {
        init(isConnected.size());

        for (int i = 0; i < isConnected.size(); i++) {
            for (int j = 0; j < isConnected[i].size(); j++) {
                if (isConnected[i][j] == 1) {
                    unionSet(i, j);
                }
            }
        }

        int count = 0;

        for (int i = 0; i < parent.size(); i++) {
            if (parent[i] == i) {
                count++;
            }
        }

        return count;
    }
};

class Solution {
    void dfs(vector<vector<int>> &isConnected, vector<int> &visited, int current) {
        visited[current] = 1;

        for (int j = 0; j < isConnected[current].size(); j++) {
            if (isConnected[current][j] == 1 && visited[j] == 0) {
                dfs(isConnected, visited, j);
            }
        }
    }

public:
    int findCircleNum(vector<vector<int>> &isConnected) {
        vector<int> visited(isConnected.size(), 0);
        int count = 0;

        for (int i = 0; i < isConnected.size(); i++) {
            if (visited[i] == 0) {
                dfs(isConnected, visited, i);
                count++;
            }
        }

        return count;
    }
};